src/main/java/fr/xephi/authme/service/BungeeService.java

```diff
@@ -1,69 +1,191 @@
+        sendAuthMeBungeecordMessage(AuthMeBungeeMessageType.LOGIN, name.toLowerCase());
+    }
+
+    public void sendLogout(String name) {
+        sendAuthMeBungeecordMessage(AuthMeBungeeMessageType.LOGOUT, name.toLowerCase());
+    }
+
+    public void sendRegister(String name) {
+        sendAuthMeBungeecordMessage(AuthMeBungeeMessageType.REGISTER, name.toLowerCase());
+    }
+
+    public void sendUnregister(String name) {
+        sendAuthMeBungeecordMessage(AuthMeBungeeMessageType.UNREGISTER, name.toLowerCase());
+    }
+
+    public void sendRefreshPassword(String name) {
+        sendAuthMeBungeecordMessage(AuthMeBungeeMessageType.REFRESH_PASSWORD, name.toLowerCase());
+    }
+
+    public void sendRefreshSession(String name) {
+        sendAuthMeBungeecordMessage(AuthMeBungeeMessageType.REFRESH_SESSION, name.toLowerCase());
+    }
+
+    public void sendRefreshQuitLoc(String name) {
+        sendAuthMeBungeecordMessage(AuthMeBungeeMessageType.REFRESH_QUITLOC, name.toLowerCase());
+    }
+
+    public void sendRefreshEmail(String name) {
+        sendAuthMeBungeecordMessage(AuthMeBungeeMessageType.REFRESH_EMAIL, name.toLowerCase());
     }
 
     @Override
-    public void reload(Settings settings) {
-        this.isEnabled = settings.getProperty(HooksSettings.BUNGEECORD);
-        this.bungeeServer = settings.getProperty(HooksSettings.BUNGEECORD_SERVER);
-        Messenger messenger = plugin.getServer().getMessenger();
-        if (!this.isEnabled) {
+    public void onPluginMessageReceived(String channel, Player player, byte[] data) {
+        if(!isEnabled) {
             return;
         }
-        if (!messenger.isOutgoingChannelRegistered(plugin, "BungeeCord")) {
-            messenger.registerOutgoingPluginChannel(plugin, "BungeeCord");
+
+        ByteArrayDataInput in = ByteStreams.newDataInput(data);
+        String subchannel = in.readUTF();
+        if(!"Authme".equals(subchannel)) {
+            return;
+        }
+
+        String type = in.readUTF();
+        switch (type) {
+            case AuthMeBungeeMessageType.UNREGISTER:
+                handleRemove(in.readUTF());
+                break;
+            case AuthMeBungeeMessageType.REFRESH_PASSWORD:
+            case AuthMeBungeeMessageType.REFRESH_QUITLOC:
+            case AuthMeBungeeMessageType.REFRESH_EMAIL:
+            case AuthMeBungeeMessageType.REFRESH:
+                handleRefresh(in.readUTF());
+                break;
+            default:
+                ConsoleLogger.debug("Received unsupported bungeecord message type! (" + type + ")");
+        }
+    }
+
+    private void handleRefresh(String name) {
+        if(!(dataSource instanceof CacheDataSource)) {
+            return;
+        }
+        CacheDataSource cacheDataSource = (CacheDataSource) dataSource;
+
+        if (cacheDataSource.getCachedAuths().getIfPresent(name) == null) {
+            return;
+        }
+        cacheDataSource.getCachedAuths().refresh(name);
```

Comment:
I would like to have interface methods for `refresh` and `invalidate` so that we can stop performing checks à la `instanceof CacheDataSource`. If you add them as empty default methods to the interface the other implementations won't even need to be touched. :)

Reply:
Nice idea!